src/pages/ProductPage/ProductPageLogic.tsx

```diff
@@ -1,119 +1,110 @@
 import { useEffect, useState } from 'react';
-import { useParams } from 'react-router-dom';
-import { itemsAll } from '../../_mocks_/itemsAll';
-import { useShoppingCart } from '../../context/ShoppingCartContext';
-import { Product } from '../../utilities/services/items.service';
+import { Product } from '../../utilities/types/ProductType';
+import { useProducts } from '../../context/ProductContext';
+import { useShoppingCart } from '../../context/CartContext';
 
-export function useProductPageLogic() {
-    // console.log('useProductPageLogic render');
-    const { increaseCartQuantity, removeFromCart } = useShoppingCart();
-    const { id } = useParams<{ id?: string }>();
+export function useProductPageLogic(id: string) {
+  // console.log('useProductPageLogic render');
+  const { products } = useProducts();
+  const { actions } = useShoppingCart();
+  const { increaseCartQuantity } = actions;
+  const [selectedSize, setSelectedSize] = useState<string | null>(null);
+  const [mainImage, setMainImage] = useState<string | null>(null);
+  const [product, setProduct] = useState<Product | null>(null);
```

Comment:
ten `const [product, setProduct] = useState<Product | null>(null);` jest całkowicie niepotrzebny.
`const product: Product | null = products.find(item => item.id.toString() === id);`
I już nie trzeba kolejnego stanu przechowywać. Właśnie to miałem na myśli mówiąc "im mniej useState/useEffect tym lepiej"

Reply:
Te problemy z rerenderami to właśnie przez te niepotrzebne useState masz :D